### ISSUE TITLE
Avoid duplicate feature normalization updates

### ISSUE THREAD

**Issue by cswinter** [entity-neural-network/incubator#25]
We'll want some generic way of normalizing input features. The best way to do this might be as an environment wrapper. There are still various details to figure out how exactly that will look like, and how the normalization will be incorporated into network architectures in a way that can be checkpointed.

**Comment by Theomat** [comment]
I'm not sure how you envision this?
Do you mean something like `InputNorm` but as an env wrapper?
Or perhaps something completely different that would already know how to normalize features?

**Comment by cswinter** [comment]
Also not entirely sure what the best thing to do is here, but @jeremysalven [made a good case for doing feature normalization in an env wrapper](https://docs.google.com/document/d/1Q87zeY7Z4u9cU0oLoH-BPQZDBQd4tHLWiEkj5YDSGw4/edit?disco=AAAAR2OxV4s).

The main issue right now is that we can end up applying feature normalization multiple times to the same datapoint. Which is fine with the current PPO implementation where we repeat each data point for several epochs, I wouldn't expect this to cause any measurable performance difference. So not a very high priority to fix now. Potentially it could cause issues with algorithms that repeat certain data points many times by systematically skewing the mean/variance towards a subset of the data.

I'm not totally sure the correct solution is an env wrapper. If we do decide we want to do this in an env wrapper, it's probably worth waiting for a more efficient batched API interface (not sure yet what exactly that should look like) so we don't need to repeat concatenating all entities.

**Comment by Theomat** [comment]
Clearly I agree with @jeremysalwen, there are two issues with doing a running mean/variance:
1. If it's in the input of the network then this "isn't compatible with any RL algorithms which do a variable number of updates per example." quoting @jeremysalwen. This can be solved my moving the normalization in an env wrapper.
2. As an environment wrapper, if you take a normalized sample `s` that you store in a replay buffer, when replaying later the normalized sample `s` that you stored does not have the same meaning wrt to the environment since normalization parameters have been updated since, this complicates the learning task of the network, until stabilization of the normalization if there is one. This leads to having a `unnormalize` and a ` normalize` method in the env wrapper so that raw features are saved in the replay then they are normalized by the wrapper when used by the network. Perhaps there's another way of solving this issue?

**Comment by cswinter** [comment]
Yeah, that's a good point, you really want to keep the original features and then apply the current normalization parameters. Otherwise, older samples will be incompatible with the network. So maybe this shouldn't live in an env wrapper after all, but be the responsibility of the training loop to handle correctly. We could have a method on the network to enable/disable updating the running mean/variance (it already skips updates when `training` is `False`) and then the PPO loop enables the updates only for the first epoch.

**Comment by Theomat** [comment]
I'm split between making it live outside of the network and inside.

- **Inside**: we have the issues mentioned by @jeremysalwen, so a `freeze_normalization` and an `unfreeze_normalization` methods should be created and used. Problem this might not look clear to the outsider who lays eyes upon the code for the first time, likely to be forgotten with non working code.
- **Outside**: the model checkpoint doesn't includes the normalization, the advantage is that it is clear in the training loop and nothing wrong happens if the lines are forgotten, it's just that you don't have normalization, so there's nothing going on under the hood except if done explicitly.

Another thing to consider is whether you want `A0.x` and `A1.x` to have the same normalization? And then the same question for `A0.x` and `B0.x`?
E.g.: a grid on which all your entities live, it seems that you're more likely to have a stable normalization if you normalize the `x` coordinate the same way for all entities.

**Comment by cswinter** [comment]
I think the default should be to do normalization since otherwise things can break horribly. Even slightly suboptimal normalization is still vastly better than no normalization, and in the majority of cases, not calling `freeze`/`unfreeze` won't really make a difference. We might also be able to do something like print a warning if the `freeze`/`unfreeze` methods are never called.

Normalizing across entities might not be worth it, I would expect the normalization parameters to become accurate fairly quickly and not benefit more from additional data points. There will also be cases where entities have some shared features but the distribution ends up being quite different and using the same normalization is detrimental. We are going to have a lot of special handling for positional encodings anway, which is probably the most common case where this might be worth it.

**Comment by cswinter** [comment]
Another way to force users to be aware of this detail is to add a `normalization_frozen` parameter with good docs and no default value to the constructor.

**Comment by Theomat** [comment]
Alright, but I'm not so sure about the new parameter, I feel like people will most of the time copy paste the code from an example which makes the additional parameter irrelevant. Perhaps just a good tutorial or example would be enough, with a comment when we call `freeze`/`unfreeze`to explain why this is needed?
Anyway to be fair, this is not an immediate concern.
